File: core/result.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "core/result.h"

namespace hbase {

Result::~Result() {}

Result::Result(const std::vector<std::shared_ptr<Cell> > &cells, bool exists, bool stale,
               bool partial)
    : exists_(exists), stale_(stale), partial_(partial), cells_(cells) {
  row_ = (cells_.size() == 0 ? "" : cells_[0]->Row());
}

Result::Result(const Result &result) {
  exists_ = result.exists_;
  stale_ = result.stale_;
  partial_ = result.partial_;
  row_ = result.row_;
  if (!result.cells_.empty()) {
    for (const auto &cell : result.cells_) {
      cells_.push_back(cell);
    }
  }
}

const std::vector<std::shared_ptr<Cell> > &Result::Cells() const { return cells_; }

std::vector<std::shared_ptr<Cell> > Result::ColumnCells(const std::string &family,
                                                        const std::string &qualifier) const {
  std::vector<std::shared_ptr<Cell> > column_cells;
  // TODO implement a BinarySearch here ?
  for (const auto &cell : cells_) {
    if (cell->Family() == family && cell->Qualifier() == qualifier) {
      column_cells.push_back(cell);
    }
  }
  return column_cells;
}

const std::shared_ptr<Cell> Result::ColumnLatestCell(const std::string &family,
                                                     const std::string &qualifier) const {
  // TODO implement a BinarySearch here ?
  for (const auto &cell : cells_) {
    // We find the latest(first) occurrence of the Cell for a given column and
    // qualifier and break
    if (cell->Family() == family && cell->Qualifier() == qualifier) {
      return cell;
    }
  }
  return nullptr;
}

optional<std::string> Result::Value(const std::string &family, const std::string &qualifier) const {
  std::shared_ptr<Cell> latest_cell(ColumnLatestCell(family, qualifier));
  if (latest_cell.get()) {
    return optional<std::string>(latest_cell->Value());
  }
  return optional<std::string>();
}

bool Result::IsEmpty() const { return cells_.empty(); }

const std::string &Result::Row() const { return row_; }

int Result::Size() const { return cells_.size(); }

ResultMap Result::Map() const {
  ResultMap result_map;
  for (const auto &cell : cells_) {
    result_map[cell->Family()][cell->Qualifier()][cell->Timestamp()] = cell->Value();
  }
  return result_map;
}

std::map<std::string, std::string> Result::FamilyMap(const std::string &family) const {
  std::map<std::string, std::string> family_map;
  if (!IsEmpty()) {
    auto result_map = Map();
    auto itr = result_map.find(family);
    if (itr == result_map.end()) {
      return family_map;
    }

    for (auto qitr = itr->second.begin(); qitr != itr->second.end(); ++qitr) {
      for (auto vitr = qitr->second.begin(); vitr != qitr->second.end(); ++vitr) {
        // We break after inserting the first value. Result.java takes only
        // the first value
        family_map[qitr->first] = vitr->second;
        break;
      }
    }
  }

  return family_map;
}

std::string Result::DebugString() const {
  std::string ret{"keyvalues="};
  if (IsEmpty()) {
    ret += "NONE";
    return ret;
  }
  ret += "{";
  bool is_first = true;
  for (const auto &cell : cells_) {
    if (is_first) {
      is_first = false;
    } else {
      ret += ", ";
    }
    ret += cell->DebugString();
  }
  ret += "}";

  return ret;
}

size_t Result::EstimatedSize() const {
  size_t s = sizeof(Result);
  s += row_.capacity();
  for (const auto c : cells_) {
    s += sizeof(std::shared_ptr<Cell>);
    s + c->EstimatedSize();
  }
  return s;
}

} /* namespace hbase */

File: hbase-shell/src/main/ruby/shell/formatter/result.rb
#
# Copyright The Apache Software Foundation
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
require 'shell/formatter/util'

module Shell
  module Formatter
    ##
    # Mixin providing helper methods for commands that print cell-scannable results
    module ResultMixin
      private def get_row_bytes(c)
        org.apache.hadoop.hbase.util.Bytes.copy(c.getRowArray, c.getRowOffset, c.getRowLength)
      end

      private def get_family_bytes(c)
        org.apache.hadoop.hbase.util.Bytes.copy(c.getFamilyArray, c.getFamilyOffset, c.getFamilyLength)
      end

      private def get_qualifier_bytes(c)
        org.apache.hadoop.hbase.util.Bytes.copy(c.getQualifierArray, c.getQualifierOffset, c.getQualifierLength)
      end

      ##
      # iterates until cell scanner has been exhausted
      #
      # @param [org.apache.hadoop.hbase.CellScanner] cell scanner or subclass like result
      # @return [Enumerator] generates Cell objects
      def cell_iterator(cell_scanner)
        Enumerator.new do |yielder|
          loop do
            cell = cell_scanner.current
            yielder.yield cell unless cell.nil?
            break unless cell_scanner.advance
          end
        end
      end

      ##
      # iteraters until result scanner has been exhausted
      #
      # @param [ResultScanner] result_scanner
      # @return [Enumerator] generates Result objects
      def result_iterator(result_scanner)
        Enumerator.new do |yielder|
          iter = result_scanner.iterator
          loop do
            break unless iter.hasNext

            yielder.yield iter.next
          end
        end
      end

      ##
      # print cells, using converters from the given table
      #
      # @param [Enumerator] cells generates Cell objects
      # @param [Hbase::Table] table used for converters
      # @param [String] converter name of the method to use from converter_class
      # @param [String] converter_class
      def print_result(cells, table, converter = nil, converter_class = nil)
        converter ||= 'toStringBinary'
        converter_class ||= 'org.apache.hadoop.hbase.util.Bytes'

        table_formatter.start_table(
            headers: %w[ROW COLUMN TIMESTAMP VALUE],
            widths: [nil, nil, ::Shell::Formatter::Util::ISO8601_WIDTH, nil]
        )

        cells.each do |c|
          # Get the family and qualifier of the cell without escaping non-printable characters. It is crucial that
          # column is constructed in this consistent way to that it can be used as a key.
          row_bytes = get_row_bytes c
          family_bytes = get_family_bytes c
          qualifier_bytes = get_qualifier_bytes c
          column = "#{family_bytes}:#{qualifier_bytes}"

          value = table.convert(column, c, converter_class, converter)

          # Use the FORMATTER to determine how column is printed
          row = table.convert_bytes(row_bytes, converter_class, converter)
          family = table.convert_bytes(family_bytes, converter_class, converter)
          qualifier = table.convert_bytes(qualifier_bytes, converter_class, converter)
          formatted_column = "#{family}:#{qualifier}"

          timestamp = ::Shell::Formatter::Util.to_iso_8601 c.getTimestamp

          table_formatter.row([row, formatted_column, timestamp, value])
        end

        table_formatter.close_table(num_rows: table_formatter.get_rows_printed)

        nil
      end
    end
  end
end
